### DISCUSSION TITLE
Self-signed certificate with self-hosted install

### DISCUSSION THREAD

**Discussion by papiru5** [sinamics/ztnet#581] [Ideas]
Hi All!
First of all - big thanks for this great software, I've test it with pleasure! One thing I can't catch from docs or examples: how can I add self-signed certificate with debian self-hosted installation? Is it possible? If I missed some docs, please, point me to, if not - show me the way, how to do it?

**Comment by tinola** [comment]
Well, I hope you didn't get banned for searching Google for how to generate your own self-signed certificate.
If you generate a self-signed certificate and have any proxy that you can configure yourself, you should be able to connect to ZTNET via https.
You're right, this case is not yet in the documentation.

**Reply by papiru5** [reply]
So, steps would be like these:
1. Make WebUI listen only on 127.0.0.1:3000 (how?)
2. Install reverse proxy like haproxy or something similar (maybe caddy will suite better?)
3. Generate self-signed certs with openssl and put it according reverse proxy docs
4. Configure reverse proxy to transfer connections to 127.0.0.1:3000 with generated certs
Am I miss something important?

P.S. I'm not a linux guru, so this maybe naive way, if so - please, correct me.

**Reply by papiru5** [reply]
I've found some discussion there: https://github.com/sinamics/ztnet/discussions/104#discussioncomment-6893579 about number 1 in my previous reply, so I try to adapt it to debian installation. Thanks for that informative style of discussion.

**Comment by papiru5** [comment]
I've figured it out, so this is my 2 cents to this great project.

So, we've just installed self-hosted standalone ztnet at debian 12, according to official documentation: [https://ztnet.network/installation/linux](url)

Our LAN ip is 192.168.0.10, our local domain name is ztnet.mydomain.lan, our ztnet web ui is at http://192.168.0.10:3000

We need to make web ui listen only on 127.0.0.1:3000, for this we edit /opt/ztnet/server.js and change line `const hostname = process.env.HOSTNAME || '0.0.0.0'
` to `const hostname = process.env.HOSTNAME || '127.0.0.1'
`

Next, we need to edit /opt/ztnet/.env to reflect our changes, according to environment variables in documentation about NEXTAUTH Configuration: [https://ztnet.network/installation/options#nextauth-configuration](url):
NEXTAUTH_URL should be https://ztnet.mydomain.lan and we add NEXTAUTH_URL_INTERNAL=http://127.0.0.1:3000 line.

At third step we install caddy according to its official documentation: [https://caddyserver.com/docs/install#debian-ubuntu-raspbian](url).
Then we need to edit our caddy configuration file (/etc/caddy/Caddyfile) to use self-signed certificate, it should look like this:  
ztnet.mydomain.lan {  
  tls internal  
  reverse_proxy 127.0.0.1:3000  
}  
and make some magic to allow caddy to add self-signed certs to local root CA with this command: `sudo HOME=~caddy caddy trust` following this comment: [https://github.com/caddyserver/caddy/issues/4248#issuecomment-882895673](url)  
The last is to restart needed services or reboot our server.
And that's it! If ztnet.mydomain.lan is resolving to 192.168.0.10 (our ztnet web ui) we can access it at https://ztnet.mydomain.lan adding exception for our self-signed cert in browser.

Hope this guide will be helpful for others and not consist of rude mistakes.

If you won't add A record for ztnet.mydomain.lan in dns, you could change NEXTAUTH_URL to https://192.168.0.10 and do the same in Caddyfile - it would work too.

**Comment by sinamics** [comment]
Thx for the write-up and applogize for not chiming in earlier, been super busy. 

>We need to make web ui listen only on 127.0.0.1:3000, for this we edit /opt/ztnet/server.js and change line const hostname = process.env.HOSTNAME || '0.0.0.0'  to const hostname = process.env.HOSTNAME || '127.0.0.1' 

Just a tip.
You can set `HOSTNAME` and `PORT` directly in .env file

**Comment by papiru5** [comment]
Thanks again for that great software! Using .env file is more convenient, agreed, I don't guess it earlier.
